### PR TITLE
Add method to delete `allocation` by `username`

### PR DESCRIPTION
**Pull request by logist322**
## Synopsis

We are currently trying to replace [Coturn](https://github.com/coturn/coturn) with this implementation. But currently it does not have alternatives to specific Coturn's functionality e.g. providing metrics, TCP fallback, external management capabilities. So we want to elaborate on that starting with exposing some management interfaces. 

Here is the Coturn's telnet protocol for reference:

```
TURN Server
Coturn-4.5.2 'dan Eider'

  ?, h, help - print this text

  quit, q, exit, bye - end CLI session

  stop, shutdown, halt - shutdown TURN Server

  pc - print configuration

  sr <realm> - set CLI session realm

  ur - unset CLI session realm

  so <origin> - set CLI session origin

  uo - unset CLI session origin

  tc <param-name> - toggle a configuration parameter
     (see pc command output for togglable param names)

  cc <param-name> <param-value> - change a configuration parameter
     (see pc command output for changeable param names)

  ps [username] - print sessions, with optional exact user match

  psp <usernamestr> - print sessions, with partial user string match

  psd <file-name> - dump ps command output into file on the TURN server system

  pu [udp|tcp|dtls|tls]- print current users

  lr - log reset

  aas ip[:port} - add an alternate server reference
  das ip[:port] - delete an alternate server reference
  atas ip[:port] - add a TLS alternate server reference
  dtas ip[:port] - delete a TLS alternate server reference

  cs <session-id> - cancel session, forcefully

```

The particular function that we need is being able to forcefully close specific allocations.

## Solution

Add the ability to communicate with the applications main loop, running in the `Server::read_loop()` method.

Add `Server.delete_allocations_by_username(&self, username: String)` function.

Also, there are some low-hanging fruits in terms of optimizations, so this pull request includes minor refactoring aimed at improving overall performance.

### CODE REVIEW

**Comment by logist322** [comment]
@tyranron for review.

File: src/allocation/allocation_manager.rs [hunk 1]
```diff
@@ -66,36 +68,71 @@ impl Manager {
...
         if let Some(a) = allocation {
-            let mut a = a.lock().await;
             if let Err(err) = a.close().await {
                 log::error!("Failed to close allocation: {}", err);
             }
         }
     }
 
+    /// Deletes the [`Allocation`] according to the specified `username`.
+    pub async fn delete_allocations_by_username(&self, name: String) {
```
Comment (tyranron) [inline src/allocation/allocation_manager.rs:105]: @logist322 why not `name: &str`?

**Review by tyranron — CHANGES_REQUESTED** [review]

File: src/allocation/mod.rs [hunk 2]
```diff
@@ -174,12 +180,12 @@ impl Allocation {
...
     // Close closes the allocation
-    pub async fn close(&mut self) -> Result<()> {
-        if self.closed {
+    pub async fn close(&self) -> Result<()> {
+        if self.closed.load(Ordering::Acquire) {
             return Err(Error::ErrClosed);
         }
 
-        self.closed = true;
+        self.closed.store(true, Ordering::Release);
```
Comment (tyranron) [inline src/allocation/mod.rs]: @logist322 with atomics you do have atomicity regarding the single operation, but you don't have atomicity with two separate operations. So, loading and then storing separately is not atomic and will race. That's why there are compare-and-swap operations on atomics.

**Review by tyranron — APPROVED** [review]

**Comment by logist322** [comment]
@rainliu request for review.

**Comment by k0nserv** [comment]
We have migrated this crate to the monorepo([webrtc-rs/webrtc](https://github.com/webrtc-rs/webrtc/)) please re-open this pull request over there.

**Pull request closed without merging**